### PR TITLE
🌱 OCM Kueue Admission Check Controller

### PR DESCRIPTION
**Pull request by z1ens**
- Add the document for setting up Multikueue environment using OCM Kueue Admission Check Controller.
- Add files for the environment set up.
- Document still needs refinement.

Update 25.08
- Fix docu.

REF: https://github.com/open-cluster-management-io/ocm/issues/369

### CODE REVIEW

File: solutions/kueue-admission-check/setup-env.sh [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+]'
+kubectl patch clustermanager cluster-manager --type=json -p='[{"op": "replace", "path": "/spec/registrationImagePullSpec", "value": "quay.io/haoqing/registration:latest"}]'
+kubectl patch clustermanager cluster-manager --type=json -p='[{"op": "replace", "path": "/spec/placementImagePullSpec", "value": "quay.io/haoqing/placement:latest"}]'
+
+echo "Install CRDs"
+kubectl create -f env/multicluster.x-k8s.io_authtokenrequests.yaml
+kubectl create -f env/multicluster.x-k8s.io_clusterprofiles.yaml
+
+echo "Install managed-serviceaccount"
+cd /path/to/managed-serviceaccount  # TODO: Replace here with your actual path.
```
Comment (haoqing0110) [inline solutions/kueue-admission-check/setup-env.sh:76]: Replace with `git clone git@github.com:open-cluster-management-io/managed-serviceaccount.git`

File: solutions/kueue-admission-check/setup-env.sh [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+echo "Install managed-serviceaccount mca"
+clusteradm create clusterset spoke
+clusteradm clusterset set spoke --clusters ${c1},${c2},${c3}
+clusteradm clusterset bind spoke --namespace default
+kubectl apply -f env/placement.yaml || true
+kubectl apply -f env/mg-sa-cma-0.6.0.yaml || true
+
+echo "Install cluster-permission"
+cd /path/to/OCM/cluster-permission  # TODO: Replace here with your actual path.
```
Comment (haoqing0110) [inline solutions/kueue-admission-check/setup-env.sh:95]: Replace with `git clone git@github.com:open-cluster-management-io/cluster-permission.git`

File: solutions/kueue-admission-check/setup-env.sh [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+make install
+make deploy
+cd -
+
+echo "Install resource-usage-collect-addon"
+cd /path/to/addon-contrib/resource-usage-collect-addon  # TODO: Replace here with your actual path.
+IMAGE_NAME=zheshen/resource-usage-collect-addon:latest make deploy
```
Comment (haoqing0110) [inline solutions/kueue-admission-check/setup-env.sh:102]: It seems the image name is hardcoded in addontemplate https://github.com/open-cluster-management-io/addon-contrib/blob/main/resource-usage-collect-addon/deploy/resources/addon-template.yaml#L53 , need to replace it with `example-addon-image`.

File: solutions/kueue-admission-check/README.md [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+
+As an admin, I want to leverage OCM's `AddonPlacementScore` for dynamic workload scheduling, so that clusters with higher GPU scores, indicating clusters with more GPU resources, are selected and converted into a [MultiKueue](https://kueue.sigs.k8s.io/docs/concepts/multikueue/), which automatically adjusts by adding or removing clusters as scores change.
+
+`placememt-demo2-2` selects clusters with the `nvidia-tesla-t4` accelerator label, and select one cluster with the highest GPU-score, indicating having more GPU resources.
+
+```yaml
+apiVersion: cluster.open-cluster-management.io/v1beta1
+kind: Placement
+metadata:
+  name: placement-sample2
```
Comment (haoqing0110) [inline solutions/kueue-admission-check/README.md:212]: The name is inconsistent with `placement-demo2-2.yaml` resource name.

File: solutions/kueue-admission-check/README.md [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+    configurations:
+      - scoreCoordinate:
+          type: AddOn
+          addOn:
+            resourceName: resource-usage-score
+            scoreName: gpuAvailable
+        weight: 1
+```
+- You can manually edit the GPU resources on the managed clusters for testing.
+```bash
```
Comment (haoqing0110) [inline solutions/kueue-admission-check/README.md:239]: Add some details on what user should add to their node.

File: solutions/kueue-admission-check/multikueue-setup-demo2.yaml [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+apiVersion: kueue.x-k8s.io/v1beta1
+kind: AdmissionCheck
+metadata:
+  name: ocm-multikueue
+spec:
+  controllerName: open-cluster-management.io/placement
+  parameters:
+    apiGroup: cluster.open-cluster-management.io
+    kind: Placement
+    name: placement-sample1  # An example placement to select clusters labeled with "nvidia-tesla-t4" GPU accelerator.
```
Comment (haoqing0110) [inline solutions/kueue-admission-check/multikueue-setup-demo2.yaml:55]: This is the placement name, should be placement-demo2.

File: solutions/kueue-admission-check/env/authtokenrequest-c1.yaml [hunk 7]
```diff
@@ -0,0 +1,66 @@
+apiVersion: multicluster.x-k8s.io/v1alpha1
+kind: AuthTokenRequest
```
Comment (qiujian16) [inline solutions/kueue-admission-check/env/authtokenrequest-c1.yaml:2]: I would not think we should have this, since this is not decided yet.

Comment (haoqing0110) [reply solutions/kueue-admission-check/env/authtokenrequest-c1.yaml:2]: yes, let's replace it with clusterpermission and managedserviceaccount.

File: solutions/kueue-admission-check/setup-env.sh (hunk 3 above)
Comment (z1ens) [reply solutions/kueue-admission-check/setup-env.sh:102]: Yes a new PR has been created to fix the image problem.
https://github.com/open-cluster-management-io/addon-contrib/pull/21

File: solutions/kueue-admission-check/env/patch-clusterrole.json [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+  },
+  {
+    "op": "add",
+    "path": "/rules/-",
+    "value": {
+      "apiGroups": ["multicluster.x-k8s.io"],
+      "resources": ["authtokenrequests/status"],
+      "verbs": ["update", "patch"]
+    }
+  },
```
Comment (haoqing0110) [inline solutions/kueue-admission-check/env/patch-clusterrole.json:37]: Remove the permission of `authtokenrequests` and `authtokenrequests/status`.

File: solutions/kueue-admission-check/README.md [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+kubectl get multikueuecluster -A -ojson | jq '.items[] | .metadata.name, .status.conditions'
+kubectl get admissionchecks -ojson | jq '.items[] | .metadata.name, .status.conditions'
+kubectl get clusterqueues -ojson | jq '.items[] | .metadata.name, .status.conditions'
+```
+- Deploy a job to the MultiKueue.
+```bash
+kubectl create -f ./job-demo1.yaml
+```
+- Check the workload on the managed clusters.
```
Comment (haoqing0110) [inline solutions/kueue-admission-check/README.md:134]: Adding the expected output for the "Check" would be nice.

File: solutions/kueue-admission-check/README.md [hunk 10]
```diff
@@ -0,0 +1,383 @@
...
+  name: multikueue-cluster2
+spec:
+  kubeConfig:
+    locationType: Secret
+    location: kueue-admin-cluster2-kubeconfig
+```
+
+#### After Using OCM Admission Check Controller
+
+With the OCM Admission Check Controller, the need for manual configuration of `MultiKueueConfig` and `MultiKueueCluster` is eliminated. Instead, the administrator only needs to configure two additional admission checks in the ClusterQueue resource:
```
Comment (haoqing0110) [inline solutions/kueue-admission-check/README.md:351]: Briefly explain why "only need to configure two additional admission checks" here would be nice.

**Comment by haoqing0110** [comment]
@qiujian16 could you help take a final review?

**Comment by qiujian16** [comment]
I think we should add some TODO and then we can merge
1. where to run admission check code
2. how to enable this in ocm.

/approve
/hold

**Pull request merged**